Fold LBOUND and UBOUND intrinsic functions

klausler opened the pull request:
Also tweak expression semantics so that it no longer inserts empty subscript triplets into whole array references (that breaks these bounds intrinsics), and restructure `GetShape` so that it uses the expression visitation framework rather than a class of mutually recursive functions.

jeanPerier commented on lib/evaluate/intrinsics.cc:
```diff
@@ -1497,8 +1497,9 @@ static bool ApplySpecificChecks(
-              symbol->attrs().test(semantics::Attr::ALLOCATABLE);
+          const Symbol &resolved{ResolveAssociations(*symbol)};
+          ok = resolved.has<semantics::ObjectEntityDetails>() &&
+              resolved.attrs().test(semantics::Attr::ALLOCATABLE);
```
Technically,   `Within an ASSOCIATE, CHANGE TEAM, or SELECT TYPE construct, [...] The associating entity does not have the ALLOCATABLE or POINTER attributes`
Which only leaves SELECT RANK where it would be legal to call on an associated entity.
However, I am not sure it is very harmful to allow the use of ALLOCATED in all associations kind.
Should you also call `ResolveAssociations` for `ASSOCIATED` below ?

klausler replied:
Good point.  I'll revert the test to a simple check for the ALLOCATABLE attribute on the original symbol, and trust name resolution to get it right in the case of SELECT RANK.

klausler commented on lib/evaluate/traversal.h:
```diff
@@ -90,7 +90,8 @@ template<typename RESULT> class VisitorBase {
   void Return() { done_ = true; }
-  void Return(RESULT &&x) {
+
+  template<typename A> void Return(A &&x) {
```
Good idea.

klausler commented on lib/evaluate/shape.cc:
```diff
@@ -371,102 +397,45 @@ std::optional<Shape> GetShapeHelper::GetShape(const CoarrayRef &coarrayRef) {
+    Scalar();
   } else if (const Symbol * symbol{call.proc().GetSymbol()}) {
-    return GetShape(*symbol);
+    Handle(*symbol);
```
It's valid, and the result will be `[29,28]` since the argument is not a whole array.

klausler replied:
FIxed and tested.

klausler commented on lib/evaluate/fold.cc:
```diff
@@ -717,13 +762,70 @@ Expr<Type<TypeCategory::Integer, KIND>> FoldIntrinsicFunction(
         return Expr<T>{ConvertToType<T>(Fold(context, std::move(product)))};
       }
     }
+  } else if (name == "ubound") {
```
Maybe, but the folding code for both of them uses utility subroutines in `fold.cc`.  I'll extract them to their own routines in `fold.cc`.

klausler commented on lib/evaluate/fold.cc:
```diff
@@ -717,13 +762,70 @@ Expr<Type<TypeCategory::Integer, KIND>> FoldIntrinsicFunction(
     }
+  } else if (name == "ubound") {
+    if (auto *array{UnwrapExpr<Expr<SomeType>>(args[0])}) {
+      if (int rank{array->Rank()}; rank > 0) {
```
I thought that the implicit comparison to zero was too obscure, so I made the test explicit for better readability.

klausler commented on lib/evaluate/shape.cc:
```diff
@@ -190,7 +192,7 @@ bool ContainsAnyImpliedDoIndex(const ExtentExpr &expr) {
 MaybeExtentExpr GetLowerBound(
     FoldingContext &context, const NamedEntity &base, int dimension) {
-  const Symbol &symbol{base.GetLastSymbol()};
+  const Symbol &symbol{ResolveAssociations(base.GetLastSymbol())};
```
Fixed, thanks.

klausler commented on lib/evaluate/fold.cc:
```diff
@@ -551,6 +550,52 @@ Expr<Type<TypeCategory::Integer, KIND>> FoldIntrinsicFunction(
+                           AsConstantShape(GetLowerBounds(context, *named))}) {
+              return Fold(context,
+                  ConvertToType<T>(Expr<ExtentType>{std::move(*lbounds)}));
+            }
```
Even better, I will call `AsExtentArrayExpr` and fold its result.  This will still produce a constant if possible.  Thanks!

klausler commented:
Thank you for the detailed examination of the changes.  I believe that I have addressed all of your comments and suggestions with the commits that I have pushed today.
